server/src/main/java/org/apache/druid/catalog/model/table/LocalInputSourceDefn.java

```diff
@@ -0,0 +1,281 @@
+          FILES_PARAMETER
+      );
+    }
+    if (hasBaseDir && hasFiles) {
+      if (hasFilter) {
+        throw new IAE(
+            "A local input source can set parameter %s or %s, but not both.",
+            FILES_PARAMETER,
+            FILTER_PARAMETER
+        );
+      }
+      // Special workaround: if the user provides a base dir, and files, convert
+      // the files to be absolute paths relative to the base dir. Then, remove
+      // the baseDir, since the Druid input source does not allow both a baseDir
+      // and a list of files.
+      jsonMap.put(FILES_FIELD, absolutePath(baseDirParam, filesParam));
+      return;
+    }
+    if (!hasBaseDir && !Strings.isNullOrEmpty(filterParam)) {
+      throw new IAE(
+          "If a local input source sets parameter %s, it must also set parameter %s",
+          FILTER_PARAMETER,
+          BASE_DIR_PARAMETER
+      );
+    }
+    if (hasBaseDir) {
+      jsonMap.put(BASE_DIR_FIELD, baseDirParam);
+    }
+    if (hasFiles) {
+      jsonMap.put(FILES_FIELD, filesParam);
+    }
+    if (filterParam != null) {
+      jsonMap.put(FILTER_FIELD, filterParam);
+    }
+  }
+
+  private List<String> absolutePath(String baseDirPath, List<String> files)
+  {
+    final File baseDir = new File(baseDirPath);
+    return files.stream()
+        .map(f -> new File(baseDir, f).toString())
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  public TableFunction partialTableFn(ResolvedExternalTable table)
+  {
+    final Map<String, Object> sourceMap = new HashMap<>(table.inputSourceMap);
+    final boolean hasFiles = !CollectionUtils.isNullOrEmpty(CatalogUtils.safeGet(sourceMap, FILES_FIELD, List.class));
+    final boolean hasFilter = !Strings.isNullOrEmpty(CatalogUtils.getString(sourceMap, FILTER_FIELD));
+    List<ParameterDefn> params = new ArrayList<>();
+    if (!hasFiles && !hasFilter) {
+      params.add(FILES_PARAM_DEFN);
+      params.add(FILTER_PARAM_DEFN);
+    }
+
+    // Does the table define a format?
+    if (table.inputFormatMap == null) {
+      params = addFormatParameters(params);
+    }
+    return new PartialTableFunction(table, params);
+  }
+
+  @Override
+  protected ExternalTableSpec convertCompletedTable(
+      final ResolvedExternalTable table,
+      final Map<String, Object> args,
+      final List<ColumnSpec> columns
+  )
+  {
+    final Map<String, Object> sourceMap = new HashMap<>(table.inputSourceMap);
+    final boolean hasFiles = !CollectionUtils.isNullOrEmpty(CatalogUtils.safeGet(sourceMap, FILES_FIELD, List.class));
+    if (hasFiles) {
+      if (!args.isEmpty()) {
+        throw new IAE("The local input source has a file list: do not provide other arguments");
+      }
+    } else {
+      final String baseDir = CatalogUtils.getString(sourceMap, BASE_DIR_FIELD);
+      if (Strings.isNullOrEmpty(baseDir)) {
+        throw new IAE(
+            "When an inline external table is used with a table function, %s must be set",
```

Comment:
why is this error message talking about inline external table? Isnt this a Local input source?